File: app/src/main/java/com/ivanovsky/passnotes/domain/FileHelper.java
package com.ivanovsky.passnotes.domain;

import android.content.Context;

import com.ivanovsky.passnotes.data.repository.SettingsRepository;
import com.ivanovsky.passnotes.util.InputOutputUtils;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.UUID;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class FileHelper {

	private final Context context;
	private final SettingsRepository settings;

	public FileHelper(Context context, SettingsRepository settings) {
		this.context = context;
		this.settings = settings;
	}

	@Nullable
	public File generateDestinationFileForRemoteFile() {
		File result = null;

		String destination = generateDestinationForRemoteFile();
		if (destination != null) {
			result = new File(destination);
		}

		return result;
	}

	@Nullable
	public String generateDestinationForRemoteFile() {
		String result = null;

		File dir = getRemoteFilesDir();
		if (dir != null) {
			result = dir.getPath() + "/" + UUID.randomUUID().toString();
		}

		return result;
	}

	@Nullable
	public File getRemoteFilesDir() {
		File result;

		if (settings.isExternalStorageCacheEnabled()) {
			result = getExternalStorageDir("remote-files");
		} else {
			result = getPrivateDir("remote-files");
		}

		return result;
	}

	@Nullable
	private File getPrivateDir(@NonNull String name) {
		File result = null;

		File dir = context.getDir(name, Context.MODE_PRIVATE);
		if (dir != null && dir.exists()) {
			result = dir;
		}

		return result;
	}

	private File getExternalStorageDir(@NonNull String name) {
		File result = null;

		File dir = context.getExternalCacheDir();
		if (dir != null && dir.exists()) {
			File subDir = new File(dir, name);
			if (subDir.exists() || (!subDir.exists() && subDir.mkdirs())) {
				result = subDir;
			}
		}

		return result;
	}

	@Nullable
	public File getDatabaseDir() {
		File result = null;

		File filesDir = context.getFilesDir();
		if (filesDir != null && filesDir.getParentFile() != null) {

			File dataDir = filesDir.getParentFile();
			if (dataDir.exists()) {

				File databaseDir = new File(dataDir.getPath() + File.separator + "databases");
				if (databaseDir.exists()) {
					result = databaseDir;
				}
			}
		}

		return result;
	}

	public boolean isLocatedInPrivateStorage(@NonNull File file) {
		boolean result = false;

		File privateDir = context.getFilesDir();
		if (privateDir != null) {
			File dataDir = privateDir.getParentFile();
			if (dataDir != null) {
				String dataDirPath = dataDir.getPath();

				result = file.getPath().startsWith(dataDirPath);
			}
		}

		return result;
	}

	@Nullable
	public void duplicateFile(@NonNull File src, @NonNull File destination)
			throws IOException {
		InputStream in = new BufferedInputStream(new FileInputStream(src));
		OutputStream out = new BufferedOutputStream(new FileOutputStream(destination));

		InputOutputUtils.copy(in, out, true);
	}
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/domain/FileHelper.kt
package com.ivanovsky.passnotes.domain

import android.content.Context
import com.ivanovsky.passnotes.data.entity.OperationError.MESSAGE_FAILED_TO_ACCESS_TO_PRIVATE_STORAGE
import com.ivanovsky.passnotes.data.entity.OperationError.MESSAGE_FAILED_TO_CREATE_A_DIRECTORY
import com.ivanovsky.passnotes.data.entity.OperationError.newGenericIOError
import com.ivanovsky.passnotes.data.entity.OperationResult
import com.ivanovsky.passnotes.data.repository.settings.Settings
import com.ivanovsky.passnotes.domain.entity.exception.Stacktrace
import java.io.File
import java.util.UUID

class FileHelper(
    private val context: Context,
    private val settings: Settings
) {

    val filesDir: File?
        get() = context.filesDir

    val remoteFilesDir: File?
        get() {
            return if (settings.isExternalStorageCacheEnabled) {
                getExternalPrivateStorageDir(REMOTE_FILES_DIR_NAME)
            } else {
                getInternalPrivateDir(REMOTE_FILES_DIR_NAME)
            }
        }

    fun generateDestinationFileOrNull(): File? {
        val path = generateDestinationForFile(Location.REMOTE_FILES) ?: return null

        return File(path)
    }

    fun generateDestinationFile(): OperationResult<File> {
        val file = generateDestinationFileOrNull()

        return if (file != null) {
            OperationResult.success(file)
        } else {
            OperationResult.error(
                newGenericIOError(
                    MESSAGE_FAILED_TO_ACCESS_TO_PRIVATE_STORAGE,
                    Stacktrace()
                )
            )
        }
    }

    fun generateDestinationDirectoryForSharedFile(): OperationResult<File> {
        val path = generateDestinationForFile(Location.SHARED_FILES)
            ?: return OperationResult.error(
                newGenericIOError(
                    MESSAGE_FAILED_TO_ACCESS_TO_PRIVATE_STORAGE,
                    Stacktrace()
                )
            )

        val dir = File(path)

        if (!dir.mkdirs()) {
            return OperationResult.error(
                newGenericIOError(
                    MESSAGE_FAILED_TO_CREATE_A_DIRECTORY,
                    Stacktrace()
                )
            )
        }

        return OperationResult.success(dir)
    }

    fun generateDestinationForPrivateFile(name: String?): OperationResult<File> {
        val path = generateDestinationForFile(
            location = Location.PRIVATE_FILES,
            baseName = name
        ) ?: return OperationResult.error(
            newGenericIOError(
                MESSAGE_FAILED_TO_ACCESS_TO_PRIVATE_STORAGE,
                Stacktrace()
            )
        )

        return OperationResult.success(File(path))
    }

    fun isLocatedInInternalStorage(file: File): Boolean {
        val dataDirPath = context.filesDir?.parentFile?.path
        return dataDirPath != null && file.path.startsWith(dataDirPath)
    }

    private fun generateDestinationForFile(
        location: Location,
        baseName: String? = null
    ): String? {
        val dir = when (location) {
            Location.REMOTE_FILES -> remoteFilesDir
            Location.SHARED_FILES -> context.cacheDir
            Location.PRIVATE_FILES -> filesDir
        } ?: return null

        return if (baseName != null) {
            val file = File(dir.path, baseName)
            if (file.exists()) {
                "${dir.path}/$baseName-${UUID.randomUUID()}"
            } else {
                file.path
            }
        } else {
            dir.path + "/" + UUID.randomUUID().toString()
        }
    }

    private fun getInternalPrivateDir(name: String): File? {
        val dir = context.getDir(name, Context.MODE_PRIVATE)
        return if (dir != null && dir.exists()) {
            dir
        } else {
            null
        }
    }

    private fun getExternalPrivateStorageDir(name: String): File? {
        val dir = context.externalCacheDir
        return if (dir != null && dir.exists()) {
            val subDir = File(dir, name)
            if (subDir.exists() || !subDir.exists() && subDir.mkdirs()) {
                subDir
            } else {
                null
            }
        } else {
            null
        }
    }

    enum class Location {
        REMOTE_FILES,
        SHARED_FILES,
        PRIVATE_FILES
    }

    companion object {
        private const val REMOTE_FILES_DIR_NAME = "remote-files"
    }
}